// swagen-core/lib/parser/index.js
'use strict';

const _ = require('lodash');
const DataTypeResolver = require('./data-type-resolver');

module.exports = class Parser {
    constructor(swagger) {
        this.swagger = this.getSwaggerObject(swagger);
        this.dataTypeResolver = new DataTypeResolver();
    }

    getSwaggerObject(swagger) {
        if (typeof swagger === 'string') {
            return JSON.parse(swagger);
        }
        if (typeof swagger === 'object') {
            return swagger;
        }
        throw new Error(`Unrecognized swagger data of type ${typeof swagger}:\n${swagger}`);
    }

    parse() {
        this.definition = {
            metadata: {},
            services: {},
            models: {},
            enums: {}
        };
        this.parseMetadata();
        this.parseModels();
        this.parseServices();
        return this.definition;
    }

    parseMetadata() {
        if (this.swagger.info) {
            this.definition.metadata.description = this.swagger.info.description;
            this.definition.metadata.title = this.swagger.info.title;
        }

        let baseUrl = this.swagger.schemes && typeof this.swagger.schemes === 'array' && this.swagger.schemes.length > 0
            ? this.swagger.schemes[0] : 'http';
        baseUrl += `://${this.swagger.host || 'localhost'}${this.swagger.basePath || ''}`;
        if (!_.endsWith(baseUrl, '/')) {
            baseUrl += '/';
        }
        this.definition.metadata.baseUrl = baseUrl;
    }

    parseModels() {
        for (let def in this.swagger.definitions) {
            if (this.swagger.definitions.hasOwnProperty(def)) {
                let model = this.parseModel(this.swagger.definitions[def], def);
                this.definition.models[def] = model;
            }
        }
    }

    parseModel(def, name) {
        let model = {};
        for (let prop in def.properties) {
            if (def.properties.hasOwnProperty(prop)) {
                model[prop] = this.dataTypeResolver.resolve(def.properties[prop]);
                this.tryAddEnum(model[prop], `${prop}_${name}`);
            }
        }
        return model;
    }

    parseServices() {
        for (let pathKey in this.swagger.paths) {
            for (let verbKey in this.swagger.paths[pathKey]) {
                let verb = this.swagger.paths[pathKey][verbKey];
                if (!verb.tags || !verb.tags.length) {
                    throw new Error(`Cannot figure out service name for following endpoint because it contains no tags\n:${verbKey} ${pathKey}`);
                }
                let serviceName = verb.tags[0];
                if (!this.definition.services[serviceName]) {
                    this.definition.services[serviceName] = {};
                }
                let operationName = verb.operationId;
                if (!operationName) {
                    throw new Error(`Cannot figure out name for following endpoint because it contains no operationId\n:${verbKey} ${pathKey}`);
                }
                this.definition.services[serviceName][operationName] = this.parseOperation(pathKey, verbKey, verb, operationName);
            }
        }
    }

    parseOperation(path, verb, details, operationName) {
        let operation = {
            description: details.summary,
            description2: details.description,
            path: path,
            verb: verb,
            parameters: []
        };

        if (details.parameters && details.parameters.length) {
            for (let i = 0; i < details.parameters.length; i++) {
                let param = details.parameters[i];
                let parameter = {
                    name: param.name,
                    type: param['in'],
                    description: param.description,
                    required: param.required,
                    dataType: this.dataTypeResolver.resolve(param)
                };
                this.tryAddEnum(parameter.dataType, `${operationName}_${param.name}`);
                operation.parameters.push(parameter);
            }
        }

        if (details.responses) {
            operation.responses = {};
            for (let statusCodeKey in details.responses) {
                let response = details.responses[statusCodeKey];
                let statusCode = +statusCodeKey;
                let dataType = response.schema ? this.dataTypeResolver.resolve(response.schema) : undefined;

                let responseDetails = {};
                if (dataType) {
                    responseDetails.dataType = dataType;
                    this.tryAddEnum(dataType, `${operationName}_Result_${statusCodeKey}`);
                }
                operation.responses[statusCodeKey] = responseDetails;
            }
        }

        return operation;
    }

    tryAddEnum(dataType, recommendedName) {
        if (!dataType.enum) {
            return;
        }
        let valueHash = dataType.values.join();
        for (let e in this.definition.enums) {
            let compareHash = this.definition.enums[e].join();
            if (valueHash === compareHash) {
                dataType.enum = e;
                delete dataType.values;
                return;
            }
        }
        dataType.enum = recommendedName;
        this.definition.enums[recommendedName] = dataType.values;
        delete dataType.values;
    }
}
