// swagen-ng1-http/index.js
'use strict';

const os = require('os');

const modeMappings = {
    typescript: 'typescript',
    ts: 'typescript',
    typescript_original: 'typescript',
    ts_original: 'typescript'
};

const modes = [
    {
        name: 'typescript',
        description: 'Generates Typescript code (deprecated)',
        language: 'typescript',
        extension: 'ts'
    }
];

function generate(definition, profile) {
    return [
        `The swagen-ng1-http generator has been deprecated and replaced by the swagen-angular package.`,
        ``,
        `To upgrade, follow these steps:`,
        `1. Install the swagen-angular package`,
        `2. Update your swagen config as follows:`,
        `   * Change generator value from 'ng1-http' to 'angular'.`,
        `   * Change mode value from 'typescript' or 'typescript_original' to 'ng1-typescript'.`,
        `3. If you are using a typed configuration (swagen.config.ts), the typings has been changed.`,
        `   You will need to update the config code accordingly.`,
        ``,
        `All other configurations remain the same.`,
        ``,
        `Please raise any issues in the swagen-angular project (https://github.com/swagen/swagen-angular/issues)`
    ].join(os.EOL);
}

module.exports = {
    modes,
    generate
};
